extensions/interactions/RatioExpressionInput/directives/ratio-expression-input-rules.service.ts

```diff
@@ -0,0 +1,45 @@
+// Copyright 2019 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Rules service for the RatioExpressionInput interaction.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { RatioInputAnswer } from 'interactions/answer-defs';
+import {
+  RatioInputRulesInputs
+} from 'interactions/rule-input-defs';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class RatioExpressionInputRulesService {
+  Equals(answer: RatioInputAnswer, inputs: RatioInputRulesInputs):
+  boolean {
+    return answer === inputs.x;
```

Comment:
So, the input is a string, right? What if the learner enters '5 : 3', will it be matched to '5:3' (i.e are spaces handled)?

Reply:
if users enter a space that won't be identified as a valid regex pattern and submit button would be disabled

Reply:
no, we don't support decimal for now

Reply:
@seanlip what are your thoughts on this

Reply:
OK not to support decimals.

But it seems reasonable to support spaces? Just trim them before comparing.

Reply:
Btw, also, please come up with a list of error messages to show, and send to me for review. @iamprayush can help you with that (he did it for his math input interactions).

Reply:
okay would do that

Reply:
@seanlip should I make a object factory to display different error messages of simple string comparison would be good enough?

Reply:
Oh, just to clarify the point above, I meant, make a spreadsheet of error messages and the conditions under which they occur (perhaps ask @iamprayush to send you his, so you have an example.

Re how to display them: look at other interactions and follow existing patterns.

Reply:
I did make the spreadsheet(sharing in few minutes) but this was regarding the existing pattern that you are mentioning above(for eg fraction - it have fraction object factory which throws the error) but in case of ratio i don't think so that I would be requiring an object factory.

I'm thinking something like this
```
        ctrl.isValidRatio = function(value) {
          if (value.length === 0) {
            errorMessage = 'Please enter a non-empty ratio value.';
            return false;
          }
          var RATIO_REGEX = /^\d+(:\d+)+$$/;
          if (!RATIO_REGEX.test(value)) {
            errorMessage =
              ObjectsDomainConstants.RATIO_PARSING_ERRORS.INVALID_FORMAT;
          } else {
            ctrl.value = value.toString();
            errorMessage = '';
          }
          return true;
        };
```

Reply:
@aks681 your thoughts on this?

Reply:
My previous comments still apply: follow whatever pattern is used by other interactions.

Reply:
> should I make a object factory to display different error messages of simple string comparison would be good enough?

I don't think you need an object factory if you just want to show warnings based on some given conditions. If the same warnings are being shown in multiple components, maybe create a service and call functions from there; if not, you can simply implement this in the validation function itself.